### PR TITLE
feat: update SDK to work with refactored delegation pallet

### PR DESCRIPTION
**Pull request by Diiaablo95**
## fixes KILTProtocol/ticket#1485

Merges `DelegationBaseNode` and `DelegationNode` into `DelegationNode`, deletes `DelegationRootNode` and adds `DelegationHierarchyDetails`.

Integration tests are passing if using the [blockchain version refactoring the delegation pallet](https://github.com/KILTprotocol/mashnet-node/pull/226) and the [latest updated types](https://github.com/KILTprotocol/type-definitions/pull/10). Of course, [some tests are broken](https://github.com/KILTprotocol/sdk-js/pull/406/checks?check_run_id=3199647850) as we already use the new DID pallet on the server.

One thing that requires more attention is how to synchronise the state of a node with what's on the blockchain. For instance, how to fetch the latest vector of children IDs if new nodes are added under a given node. For the moment, I have added a utility function `core/src/delegation/DelegationNode.utils.ts` with the signature `export async function getSyncedState(delegationNode: DelegationNode): Promise<DelegationNode>` to update and return a node state. We might want to think about some other ways to do that.

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

File: packages/core/src/delegation/DelegationNode.ts
Comment (rflechtner) [inline packages/core/src/delegation/DelegationNode.ts:256]: I think your in-editor linting may be fooling you here, this should be understood by typescript
```suggestion
      return storeAsDelegation(this, signature)
```

File: packages/core/src/delegation/DelegationNode.ts
Comment (rflechtner) [inline packages/core/src/delegation/DelegationNode.ts:171]: To do this, something like a `.refetch()` or `.refresh()` method would be cool on the delegation node I think, returning a freshly queried version of the current node.

File: packages/core/src/delegation/DelegationNode.ts
Comment (rflechtner) [inline packages/core/src/delegation/DelegationNode.ts:308]: what do you think about refetching here to make sure we are up to date?

File: packages/core/src/delegation/DelegationNode.utils.ts
Comment (rflechtner) [inline packages/core/src/delegation/DelegationNode.utils.ts:83]:
```suggestion
  delegationNode: IDelegationNode
```

**Review by rflechtner — APPROVED** [review]
A bunch of suggestions but nothing critical. Nice work!

File: packages/core/src/delegation/DelegationDecoder.ts
Comment (rflechtner) [reply packages/core/src/delegation/DelegationDecoder.ts:21]: it is. nevermind

File: packages/core/src/delegation/DelegationDecoder.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationDecoder.ts:93]: And also on the mashnet node. I would simply leave it for now, honestly.

Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationDecoder.ts:93]: At the end it is not semantically wrong, as the hierarchy ID === root node ID.

File: packages/core/src/delegation/DelegationDecoder.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationDecoder.ts:103]: Oh yes, sorry. I changed it but I must have fu***d something up afterwards.

File: packages/core/src/delegation/DelegationNode.utils.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationNode.utils.ts:83]: We don't have this at all anymore. Resolving it then.

File: packages/core/src/delegation/DelegationDecoder.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationDecoder.ts]: Fixed in https://github.com/KILTprotocol/sdk-js/pull/406/commits/8a20ab281b2ba7d57cd26af3a0736060d6fa3309.

File: packages/core/src/delegation/DelegationDecoder.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationDecoder.ts]: Fixed in https://github.com/KILTprotocol/sdk-js/pull/406/commits/8a20ab281b2ba7d57cd26af3a0736060d6fa3309.

File: packages/core/src/delegation/DelegationDecoder.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationDecoder.ts]: Fixed in https://github.com/KILTprotocol/sdk-js/pull/406/commits/8a20ab281b2ba7d57cd26af3a0736060d6fa3309.

File: packages/core/src/delegation/DelegationNode.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationNode.ts]: Fixed in https://github.com/KILTprotocol/sdk-js/pull/406/commits/8a20ab281b2ba7d57cd26af3a0736060d6fa3309.

File: packages/core/src/delegation/DelegationNode.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationNode.ts]: Fixed in https://github.com/KILTprotocol/sdk-js/pull/406/commits/8a20ab281b2ba7d57cd26af3a0736060d6fa3309.

File: packages/core/src/delegation/DelegationNode.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationNode.ts]: Fixed in https://github.com/KILTprotocol/sdk-js/pull/406/commits/8a20ab281b2ba7d57cd26af3a0736060d6fa3309.

File: packages/core/src/delegation/DelegationNode.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationNode.ts:320]: Fixed in https://github.com/KILTprotocol/sdk-js/pull/406/commits/8a20ab281b2ba7d57cd26af3a0736060d6fa3309.

**Comment by Diiaablo95** [comment]
@rflechtner good catches, thanks! I am not sure how to proceed now, as test cases are broken. Should we merge it? Should we not merge it?

File: packages/core/src/delegation/DelegationDecoder.ts
Comment (rflechtner) [reply packages/core/src/delegation/DelegationDecoder.ts:82]: true, we could describe it as (hierarchy/root node) id. In fact, technically, you wouldn't have to mess with the mashnet node though, because we can choose field names at will with the types, and they don't have to match anything on the mashnet implementation. The mapping is purely positional, just the type needs to be right.

**Comment by rflechtner** [comment]
> @rflechtner good catches, thanks! I am not sure how to proceed now, as test cases are broken. Should we merge it? Should we not merge it?

I say merge it - the delegationNode tests succeed and that's enough for me. Tests are broken on develop and need fixing after the DID stuff is merged.
We should however first address the issue concerning the `IDelegationHierarchyDetails` interface before we merge:
https://github.com/KILTprotocol/sdk-js/pull/406#discussion_r679843277

File: packages/core/src/delegation/DelegationNode.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationNode.ts:13]: That's an old one, I'll delete it.

File: packages/core/src/delegation/DelegationNode.chain.ts
Comment (Diiaablo95) [reply packages/core/src/delegation/DelegationNode.chain.ts:150]: I cannot get the compiler to like it...

**Comment by Diiaablo95** [comment]
@rflechtner please take one last look at the changes. I found a bug in the `refreshState()` function which did not await for the data to be fetched from the chain. I also fixed the unit and integration tests which are now working again (for delegations).

**Comment by rflechtner** [comment]
> @rflechtner please take one last look at the changes. I found a bug in the `refreshState()` function which did not await for the data to be fetched from the chain. I also fixed the unit and integration tests which are now working again (for delegations).

Looks good to me! Especially after the Object.assign() issue was solved

**Pull request merged**
